### **Backpropagation From Scratch - Regression**

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.DataFrame([[8,8,4],[7,9,5],[6,10,6],[5,12,7]], columns=['cgpa', 'profile_score', 'lpa'])

In [3]:
df

cgpa  profile_score  lpa
0     8              8    4
1     7              9    5
2     6             10    6
3     5             12    7

In [4]:
# Initializing weights and biases

def initialize_parameters(layer_dims):

  np.random.seed(3)
  parameters = {}
  L = len(layer_dims)

  for l in range(1, L):

    parameters['W' + str(l)] = np.ones((layer_dims[l-1], layer_dims[l]))*0.1
    parameters['b' + str(l)] = np.zeros((layer_dims[l], 1))


  return parameters

In [16]:
initialize_parameters([2,2,1])

{'W1': array([[0.1, 0.1],
        [0.1, 0.1]]),
 'b1': array([[0.],
        [0.]]),
 'W2': array([[0.1],
        [0.1]]),
 'b2': array([[0.]])}

In [5]:
# Activation function (linear)

def linear_forward(A_prev, W, b):

  Z = np.dot(W.T, A_prev) + b

  return Z

In [9]:
# Forward Propagation

def L_layer_forward(X, parameters):

  A = X
  L = len(parameters) // 2                  # number of layers in the neural network

  for l in range(1, L+1):
    A_prev = A
    Wl = parameters['W' + str(l)]
    bl = parameters['b' + str(l)]
    #print("A"+str(l-1)+": ", A_prev)
    #print("W"+str(l)+": ", Wl)
    #print("b"+str(l)+": ", bl)
    #print("--"*20)

    A = linear_forward(A_prev, Wl, bl)
    #print("A"+str(l)+": ", A)
    #print("**"*20)

  return A,A_prev

In [7]:
# Updating weights and biases

def update_parameters(parameters,y,y_hat,A1,X):
  parameters['W2'][0][0] = parameters['W2'][0][0] + (0.001 * 2 * (y - y_hat)*A1[0][0])
  parameters['W2'][1][0] = parameters['W2'][1][0] + (0.001 * 2 * (y - y_hat)*A1[1][0])
  parameters['b2'][0][0] = parameters['W2'][1][0] + (0.001 * 2 * (y - y_hat))

  parameters['W1'][0][0] = parameters['W1'][0][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0]*X[0][0])
  parameters['W1'][0][1] = parameters['W1'][0][1] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0]*X[1][0])
  parameters['b1'][0][0] = parameters['b1'][0][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0])

  parameters['W1'][1][0] = parameters['W1'][1][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0]*X[0][0])
  parameters['W1'][1][1] = parameters['W1'][1][1] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0]*X[1][0])
  parameters['b1'][1][0] = parameters['b1'][1][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0])

In [13]:
# epochs implementation
# Parameter initialization
parameters = initialize_parameters([2,2,1])
epochs = 10

for i in range(epochs):

  Loss = []

  for j in range(df.shape[0]):

    X = df[['cgpa', 'profile_score']].values[j].reshape(2,1) # Shape(no of features, no. of training example)
    y = df[['lpa']].values[j][0]

    y_hat,A1 = L_layer_forward(X,parameters)
    y_hat = y_hat[0][0]

    update_parameters(parameters,y,y_hat,A1,X)

    Loss.append((y-y_hat)**2)

  print('Epoch - ',i+1,'Loss - ',np.array(Loss).mean())

parameters

Epoch -  1 Loss -  25.321744156025517
Epoch -  2 Loss -  18.320004165722047
Epoch -  3 Loss -  9.473661050729628
Epoch -  4 Loss -  3.2520938634031613
Epoch -  5 Loss -  1.3407132589299962
Epoch -  6 Loss -  1.1726178458115697
Epoch -  7 Loss -  1.2331289135532644
Epoch -  8 Loss -  1.2686856723180788
Epoch -  9 Loss -  1.282007891421992
Epoch -  10 Loss -  1.2866806711744867


{'W1': array([[0.251385  , 0.42030991],
        [0.26250745, 0.45149331]]),
 'b1': array([[0.02847052],
        [0.03098528]]),
 'W2': array([[0.42075476],
        [0.49891207]]),
 'b2': array([[0.50135566]])}

### **Backpropagation using keras library**

In [14]:
import tensorflow
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense

In [25]:
model = Sequential()

model.add(Dense(2, activation= 'linear', input_dim = 2))
model.add(Dense(1, activation = 'linear'))

In [26]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                      │ (None, 2)                   │               6 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 1)                   │               3 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 9 (36.00 B)

 Trainable params: 9 (36.00 B)

 Non-trainable params: 0 (0.00 B)

In [18]:
model.get_weights()

[array([[-0.8517577 ,  1.1900104 ],
        [ 0.70993733, -1.1777364 ]], dtype=float32),
 array([0., 0.], dtype=float32),
 array([[0.13347876],
        [1.2043272 ]], dtype=float32),
 array([0.], dtype=float32)]

In [19]:
# Setting weights to 0.1 and bias to 0

new_weights = [np.array([[0.1 ,  0.1 ],
        [ 0.1 , 0.1]], dtype=np.float32),
 np.array([0., 0.], dtype=np.float32),
 np.array([[ 0.1],
        [0.1]], dtype=np.float32),
 np.array([0.], dtype=np.float32)]

In [27]:
model.set_weights(new_weights)

In [28]:
model.get_weights()

[array([[0.1, 0.1],
        [0.1, 0.1]], dtype=float32),
 array([0., 0.], dtype=float32),
 array([[0.1],
        [0.1]], dtype=float32),
 array([0.], dtype=float32)]

In [29]:
optimizer = keras.optimizers.Adam(learning_rate = 0.001)
model.compile(loss = 'mean_squared_error', optimizer = optimizer)

In [30]:
model.fit(df.iloc[:,0:-1].values,df['lpa'].values, epochs= 75, verbose = 1, batch_size = 1)

Epoch 1/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - loss: 24.3749
Epoch 2/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 33.3284  
Epoch 3/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 33.5105 
Epoch 4/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 28.5523  
Epoch 5/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 25.3886  
Epoch 6/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 25.8195 
Epoch 7/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 25.4522 
Epoch 8/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 22.1099 
Epoch 9/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 26.6576  
Epoch 10/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 23.5509  
Epoch 11/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 30.1405 
Epoch 12/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 22.7365 
Epoch 13/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 26.9715 
Epoch 14/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 19.8653  
Epoch 15/75
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 17.9035 
Epoch 16/75
4/

In [31]:
model.get_weights()

[array([[0.3736347, 0.3736347],
        [0.3656057, 0.3656057]], dtype=float32),
 array([0.27231398, 0.27231398], dtype=float32),
 array([[0.37291583],
        [0.37291583]], dtype=float32),
 array([0.20472991], dtype=float32)]

### **Conclusion :- We can see that our algorithm has a loss of 1.28 after 10 epochs and algorithm from keras library has a loss of 1.06 after 75 epochs, So our results are similar**